integrations/access/opsgenie/app.go

```diff
@@ -17,15 +17,575 @@ limitations under the License.
+		return nil
+	}
+
+	// First, try to create a notification alert.
+	isNew, notifyErr := a.tryNotifyService(ctx, req)
+	notifyErr = trace.Wrap(notifyErr)
+
+	// To minimize the count of auto-approval tries, lets attempt it only when we just created an alert.
+	// But if there's an error, we can't really know is the alert new or not so lets just try.
+	if !isNew && notifyErr == nil {
+		return nil
+	}
+	// Don't show the error if the annotation is just missing.
+	if trace.Unwrap(notifyErr) == errSkip {
+		notifyErr = nil
+	}
+
+	// Then, try to approve the request if user is currently on-call.
+	approveErr := trace.Wrap(a.tryApproveRequest(ctx, req))
+	return trace.NewAggregate(notifyErr, approveErr)
+}
+
+func (a *App) onResolvedRequest(ctx context.Context, req types.AccessRequest) error {
+	var notifyErr error
+	if err := a.postReviewNotes(ctx, req.GetName(), req.GetReviews()); err != nil {
+		notifyErr = trace.Wrap(err)
+	}
+
+	resolution := Resolution{Reason: req.GetResolveReason()}
+	switch req.GetState() {
+	case types.RequestState_APPROVED:
+		resolution.Tag = ResolvedApproved
+	case types.RequestState_DENIED:
+		resolution.Tag = ResolvedDenied
+	}
+	err := trace.Wrap(a.resolveAlert(ctx, req.GetName(), resolution))
+	return trace.NewAggregate(notifyErr, err)
+}
+
+func (a *App) onDeletedRequest(ctx context.Context, reqID string) error {
+	return a.resolveAlert(ctx, reqID, Resolution{Tag: ResolvedExpired})
+}
+
+func (a *App) getNotifyServiceNames(req types.AccessRequest) ([]string, error) {
+	services, ok := req.GetResolveAnnotations()[types.TeleportNamespace+types.ReqAnnotationNotifyServicesLabel]
+	if !ok {
+		return nil, trace.NotFound("notify services not specified")
+	}
+	return services, nil
+}
+
+func (a *App) getOnCallServiceNames(req types.AccessRequest) ([]string, error) {
+	services, ok := req.GetResolveAnnotations()[types.TeleportNamespace+types.ReqAnnotationSchedulesLabel]
+	if !ok {
+		return nil, trace.NotFound("on-call schedules not specified")
+	}
+	return services, nil
+}
+
+func (a *App) tryNotifyService(ctx context.Context, req types.AccessRequest) (bool, error) {
+	log := logger.Get(ctx)
+
+	serviceNames, err := a.getNotifyServiceNames(req)
+	if err != nil {
+		log.Debugf("Skipping the notification: %s", err)
+		return false, trace.Wrap(errSkip)
+	}
+
+	reqID := req.GetName()
+	reqData := RequestData{
+		User:          req.GetUser(),
+		Roles:         req.GetRoles(),
+		Created:       req.GetCreationTime(),
+		RequestReason: req.GetRequestReason(),
+	}
+
+	// Create plugin data if it didn't exist before.
+	isNew, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		if existing != nil {
+			return PluginData{}, false
+		}
+		return PluginData{RequestData: reqData}, true
+	})
+	if err != nil {
+		return isNew, trace.Wrap(err, "updating plugin data")
+	}
+
+	if isNew {
+		for _, serviceName := range serviceNames {
+			ctx, _ = logger.WithField(ctx, "opsgenie_service_name", serviceName)
```

Comment:
Shouldn't this `ctx` be shadowed ? Otherwise the context passed into `postReviewNotes` below will include a logger field containing `opsgenie_service_name` of the last service in this loop.

Reply:
I'd be explicitly calling it `alertCtx` or something, but your mileage may vary.